nicegui/storage.py

```diff
@@ -149,10 +150,21 @@ def general(self) -> Dict:
         """General storage shared between all users that is persisted on the server (where NiceGUI is executed)."""
         return self._general
 
+    @property
+    def session(self) -> Dict:
+        """Volatile client storage that is persisted on the server (where NiceGUI is
+        executed) on a per client/per connection basis.
+        Note that this kind of storage can only be used in single page applications
```

Comment:
I wonder if we could implement something where we combine your approach here with the browser Session Storage to keep the storage as long as the browser tab is not closed or domain is switched (see #2378).

Reply:
It is possible by using the browser sessionStorage as you already discussed in 2378:

* When NiceGUI loads in the browser it checks if NICE_GUY_SESSION_GUID is stored in the sessionStorage, if not in generates one
* Upon connecting (and before the page is allowed to build) the GUID is sent from the browser to the client and would be searched in some kind of auto-garbage collected global dictionary, lets name it SessionDataBackup. If it were found it will be attached to Client.state
* When the WebSocket connection completely breaks the Client.state object is added to the SessionDataBackup.
* An async running garbage collector regularly checks the timeouts of the elements in the SessionDataBackup and removes outdates ones, e.g. after x minutes.

Requires of course that trying to access to the client.state variable only to be accessible when this hand-shake was done. This way you could transfer living/non-serializable objects from one page to another, even between full-reloads.

I think though that this should be an optional feature and should be part of another PR in the future. In streamlit if I press the reload button I get a fresh page / session... as I would assume it.

So do I when I click a link in my favorites (more by random than coincident, not thinking about that I am still in the same tab actually) and instead of a "fresh" app instance I already get an active one when keeping the session alive between real page changes. I would have to check how GDPR applies here but I think having this as default behavior would also force one to add a cookie info as well.

The only reason to implement this were if one would by force try to cling to the actual, full page reload approach between page changes. And having recently attached the SPA to the existing examples, e.g. examples/modules its like day and night when switching pages (like quite visible 0.2 seconds thus visible JavaScript load times) to like zero.

Reply:
> I think though that this should be an optional feature 

I agree.

> and should be part of another PR in the future. 

Maybe. But we need to thing about how we would do it here, because it may impact the API (see below).

> In streamlit if I press the reload button I get a fresh page / session... as I would assume it.

That heavily depends on the use case. For example if you press refresh on a GitHub page, you do not end up with a new session (even if you are not logged in). Maybe we need `app.storage.tab` for a data storage which uses a key from the browsers Session Storage as you suggested and `app.storage.page` or `app.storage.client` which is dropped as soon as the Client is disconnected (eg. page reload or similar).

Reply:
Yes, really a question of the use case and I think we should let this control remain in the hands of the NiceGUI dev.

You are right, when I refresh here my page on GitHub even the text here in this text box I'm typing right now in is **preserved**. If I open another tab **it is not**. So very likely using the sessionStorage mechanism, a potential future **app.storage.tab**. And this is good of course in this case b/c imagine having typed a lot of text and by accident tapping back button the mouse, e.g. default side button on the Logitech ones, fat fingered onto reload etc., you leave the page and all text is gone.

On the other hand there are of course all the scenarios I already mentioned where this might not be desired, can't go in detail here on a public discussion.

My take away for today: I will rename **app.storage.session** to **app.storage.client** so its transparent to the user it is bound to the client (and thus *volatile*) and then its up to him/her where he/she makes use of it.